src/work_queue.cpp

```diff
@@ -3,83 +3,157 @@
 
 #include <string.h>
 #include <chrono>
+#include <cassert>
 
-WorkQueue::WorkQueue(int num_elements, int size_of_elm): 
-  len(num_elements), elm_size(size_of_elm) {
-  head = 0;
-  tail = 0;
-  no_block = false;
+WorkQueue::WorkQueue(int q_len, int max_elm_size) : len(q_len), max_elm_size(max_elm_size) {
+  non_block = false;
 
-  // malloc the memory for the work queue
-  queue_array = (queue_elm *) malloc(sizeof(queue_elm) * len);
-  data_array = (char *) malloc(elm_size * len * sizeof(char));
+  // place all nodes of linked list in the producer queue and reserve
+  // memory for the vectors
   for (int i = 0; i < len; i++) {
-    queue_array[i].data    = data_array + (elm_size * i);
-    queue_array[i].dirty   = false;
-    queue_array[i].touched = false;
-    queue_array[i].size    = 0;
+    DataNode *node = new DataNode(max_elm_size); // create and reserve space for updates
+    node->next = producer_list; // next of node is head
+    producer_list = node; // set head to new node
   }
+  consumer_list_size = 0;
 
-  printf("WQ: created work queue with %i elements each of size %i\n", len, elm_size);
+  printf("WQ: created work queue with %i elements each of size %i\n", len, max_elm_size);
 }
 
 WorkQueue::~WorkQueue() {
-  // free the queue
-  free(data_array);
-  free(queue_array);
+  // free data from the queues
+  while (producer_list != nullptr) {
+    DataNode *temp = producer_list;
+    producer_list = producer_list->next;
+    delete temp;
+  }
+  while (consumer_list != nullptr) {
+    DataNode *temp = consumer_list;
+    consumer_list = consumer_list->next;
+    delete temp;
+  }
 }
```

Comment:
There's a race condition here since we don't grab the locks. We should figure out if we care.